src/platform/macos/input.cpp

```diff
@@ -282,16 +366,99 @@ void unicode(input_t &input, char *utf8, int size) {
 }
 
 int alloc_gamepad(input_t &input, int nr, rumble_queue_t rumble_queue) {
-  BOOST_LOG(info) << "alloc_gamepad: Gamepad not yet implemented for MacOS."sv;
-  return -1;
+  // Get a reference to the IOService
+  kern_return_t ret = IOServiceGetMatchingServices(kIOMasterPortDefault, IOServiceMatching(SERVICE_NAME), &virtgamepad_iterator);
+
+  if(ret != KERN_SUCCESS) {
+    BOOST_LOG(info) << "Gamepad: Unable to access IOService."sv;
+    return 1;
+  }
+
+  // Iterate till success
+  int found = 0;
+  while((virtgamepad_service = IOIteratorNext(virtgamepad_iterator)) != IO_OBJECT_NULL) {
+    ret = IOServiceOpen(virtgamepad_service, mach_task_self(), 0, &virtgamepad_connect);
+    if(ret == KERN_SUCCESS) {
+      found = 1;
+      break;
+    }
+
+    IOObjectRelease(virtgamepad_service);
+  }
+
+  IOObjectRelease(virtgamepad_iterator);
+
+  if(!found) {
+    BOOST_LOG(info) << "Gamepad: Unable to open IOService."sv;
+    return 1;
+  }
+
+  // Fill up the input arguments.
+  virtgamepad_input[0] = (uint64_t)strdup(VIRTGAMEPAD_NAME);   // device name
```

Comment:
It looks like your driver is identifying the gamepads by name. Can we concatenate the `nr` parameter to the hardcoded name or something to support multiple gamepads?

Reply:
This name can be changed, it's dynamic.